rcl/src/rcl/log_level.c

```diff
@@ -0,0 +1,166 @@
+      }
+      return NULL;
+    }
+  }
+
+  return log_levels;
+}
+
+rcl_ret_t
+rcl_log_levels_copy(const rcl_log_levels_t * src, rcl_log_levels_t * dst)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(src, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(dst, RCL_RET_INVALID_ARGUMENT);
+  const rcl_allocator_t * allocator = &src->allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+
+  dst->logger_settings = allocator->allocate(
+    sizeof(rcl_logger_setting_t) * (src->num_logger_settings), allocator->state);
+  if (NULL == dst->logger_settings) {
+    RCL_SET_ERROR_MSG("Error allocating memory");
+    return RCL_RET_BAD_ALLOC;
+  }
+
+  dst->default_logger_level = src->default_logger_level;
+  dst->num_logger_settings = src->num_logger_settings;
+  dst->capacity_logger_settings = src->capacity_logger_settings;
+  dst->allocator = src->allocator;
+  for (size_t i = 0; i < src->num_logger_settings; ++i) {
+    dst->logger_settings[i].name =
+      rcutils_strdup(src->logger_settings[i].name, *allocator);
+    if (NULL == dst->logger_settings[i].name) {
+      if (RCL_RET_OK != rcl_log_levels_fini(dst)) {
+        RCL_SET_ERROR_MSG("Error while finalizing log levels due to another error");
+      }
+      return RCL_RET_BAD_ALLOC;
+    }
+    dst->logger_settings[i].level =
+      src->logger_settings[i].level;
+  }
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_log_levels_fini(rcl_log_levels_t * log_levels)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(log_levels, RCL_RET_INVALID_ARGUMENT);
+  const rcl_allocator_t * allocator = &log_levels->allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+  if (log_levels->logger_settings) {
+    for (size_t i = 0; i < log_levels->num_logger_settings; ++i) {
+      allocator->deallocate((void *)log_levels->logger_settings[i].name, allocator->state);
+    }
+    log_levels->num_logger_settings = 0;
+
+    allocator->deallocate(log_levels->logger_settings, allocator->state);
+    log_levels->logger_settings = NULL;
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_log_levels_shrink_to_size(rcl_log_levels_t * log_levels)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(log_levels, RCL_RET_INVALID_ARGUMENT);
+  rcl_allocator_t * allocator = &log_levels->allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+  if (0U == log_levels->num_logger_settings) {
+    allocator->deallocate(log_levels->logger_settings, allocator->state);
+    log_levels->logger_settings = NULL;
+    log_levels->capacity_logger_settings = 0;
+  } else if (log_levels->num_logger_settings < log_levels->capacity_logger_settings) {
+    log_levels->logger_settings = rcutils_reallocf(
+      log_levels->logger_settings,
+      sizeof(rcl_logger_setting_t) * log_levels->num_logger_settings, allocator);
+    if (NULL == log_levels->logger_settings) {
+      return RCL_RET_BAD_ALLOC;
+    }
```

Comment:
I don't think that `rcl_log_levels_shrink_to_size` should automatically deallocate the memory on error.
Instead, the old settings should still be valid, and the user call fini if desired.

```suggestion
    new_logger_settings = allocator->reallocate(
      log_levels->logger_settings,
      sizeof(rcl_logger_setting_t) * log_levels->num_logger_settings,
      allocator->state);
    if (NULL == new_logger_settings) {
      return RCL_RET_BAD_ALLOC;
    }
    log_levels->logger_settings = new_logger_settings;
```

Reply:
Updated